### PR TITLE
Add EC2 instance status collector

### PR DESCRIPTION
**Pull request by marians**
Towards https://github.com/giantswarm/giantswarm/issues/3609

This adds a collector for EC2 instance status.

### Preview

```
# HELP aws_operator_ec2_instance_status Gauge indicating the status of an EC2 instance. 1 = healthy, 0 = unhealthy
# TYPE aws_operator_ec2_instance_status gauge
aws_operator_ec2_instance_status{account="084190472784",cluster_id="",ec2_instance="i-00e1980a01ffcbd5f",installation="gauss",organization="",state="running",status="ok"} 1
aws_operator_ec2_instance_status{account="084190472784",cluster_id="",ec2_instance="i-03edfeed8f2c57716",installation="gauss",organization="",state="running",status="ok"} 1
aws_operator_ec2_instance_status{account="084190472784",cluster_id="",ec2_instance="i-074305fa5a2232792",installation="gauss",organization="",state="running",status="ok"} 1
aws_operator_ec2_instance_status{account="084190472784",cluster_id="",ec2_instance="i-07455bcc3831d7f2e",installation="gauss",organization="",state="running",status="ok"} 1
aws_operator_ec2_instance_status{account="084190472784",cluster_id="",ec2_instance="i-0bf169dbe1a8ace98",installation="gauss",organization="",state="running",status="ok"} 1
aws_operator_ec2_instance_status{account="084190472784",cluster_id="",ec2_instance="i-0e122c9d9185a086f",installation="gauss",organization="",state="running",status="ok"} 1
aws_operator_ec2_instance_status{account="084190472784",cluster_id="",ec2_instance="i-0fa86134a52efbe8f",installation="gauss",organization="",state="running",status="ok"} 1
```

The gauge value is 1 when `state="running" AND status="ok"`, otherwise 0.

### CODE REVIEW

File: service/collector/ec2instances.go [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+		}
+
+		up := 0
+		if statuses.InstanceState.Name != nil {
+			state = strings.ToLower(*statuses.InstanceState.Name)
+		}
+		if statuses.InstanceStatus.Status != nil {
+			status = strings.ToLower(*statuses.InstanceStatus.Status)
+		}
+		if state == "running" && status == "ok" {
```
Comment (JosephSalisbury) [inline service/collector/ec2instances.go]: I don't get why we have these labels. Wouldn't this be the value of the metric?

Comment (JosephSalisbury) [reply service/collector/ec2instances.go]: e.g:
i'd expect
```
aws_operator_ec2_status{account="270935918670",cluster_id="3tiea",ec2instance="i-03b17014ebbe7deb8",installation="ginger",organization="giantswarm"} 1
```
for a good boy
and
```
aws_operator_ec2_status{account="270935918670",cluster_id="3tiea",ec2instance="i-03b17014ebbe7deb8",installation="ginger",organization="giantswarm"} 0
```
for a bad boy (i'm assuming i'm missing a nuance here :D )

File: service/collector/ec2instances.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+	// labelInstanceStatus is a label that will contain the instance status check value
+	labelInstanceStatus = "status"
+
+	// subsystemEC2 will become the second part of the metric name, right after namespace.
+	subsystemEC2 = "ec2"
+)
+
+var (
+	ec2InstanceStatus = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, subsystemEC2, "status"),
```
Comment (xh3b4sd) [inline service/collector/ec2instances.go]: From a metrics design point of view there is more to EC2 statuses. When this is about instances, we should extend the metric name accordingly. 

```suggestion
		prometheus.BuildFQName(namespace, subsystemEC2, "instances_status"),
```

File: service/collector/ec2instances.go [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+// EC2InstancesConfig is this collector's configuration struct.
+type EC2InstancesConfig struct {
+	Helper *helper
+	Logger micrologger.Logger
+
+	InstallationName string
+}
+
+// EC2Instances is the main struct for this collector.
+type EC2Instances struct {
```
Comment (xh3b4sd) [inline service/collector/ec2instances.go]: The type suggests the filename `ec2_instances.go` to be consistent.

File: service/collector/ec2instances.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+	clientaws "github.com/giantswarm/aws-operator/client/aws"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	// labelInstance is the metric's label key that will hold the ec2 instance ID.
+	labelInstance = "ec2instance"
```
Comment (xh3b4sd) [inline service/collector/ec2instances.go]:
```suggestion
	labelInstance = "ec2_instance"
```

File: service/collector/ec2instances.go [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+	// labelInstanceStatus is a label that will contain the instance status check value
+	labelInstanceStatus = "status"
+
+	// subsystemEC2 will become the second part of the metric name, right after namespace.
+	subsystemEC2 = "ec2"
+)
+
+var (
+	ec2InstanceStatus = prometheus.NewDesc(
```
Comment (xh3b4sd) [inline service/collector/ec2instances.go]:
```suggestion
	ec2InstancesStatus = prometheus.NewDesc(
```

File: service/collector/ec2instances.go [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+// Describe emits the description for the metrics collected here.
+func (e *EC2Instances) Describe(ch chan<- *prometheus.Desc) error {
+	ch <- ec2InstanceStatus
```
Comment (xh3b4sd) [inline service/collector/ec2instances.go]:
```suggestion
	ch <- ec2InstancesStatus
```

File: service/collector/ec2instances.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+			if o.NextToken == nil {
+				break
+			}
+			input.SetNextToken(*o.NextToken)
+		}
+	}
+
+	// Iterate over found instances and emit metrics.
+	for instanceID := range instances {
+
```
Comment (xh3b4sd) [inline service/collector/ec2instances.go]:
```suggestion
```

File: service/collector/ec2instances.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+		}
+		if statuses.InstanceStatus.Status != nil {
+			status = strings.ToLower(*statuses.InstanceStatus.Status)
+		}
+		if state == "running" && status == "ok" {
+			up = 1
+		}
+
+		ch <- prometheus.MustNewConstMetric(
+			ec2InstanceStatus,
```
Comment (xh3b4sd) [inline service/collector/ec2instances.go]:
```suggestion
			ec2InstancesStatus,
```

File: service/collector/ec2instances.go (hunk 1 above)
Comment (marians) [reply service/collector/ec2instances.go]: As I said, the value is only 1 when state is "running" AND status is "ok". That's a healthy, operating instance.

An instance that has failed healthchecks will have a `status` value != `ok`. Likely the state string appearing will be `impaired`.

File: service/collector/ec2instances.go (hunk 2 above)
Comment (marians) [reply service/collector/ec2instances.go]: Yes, fine with me. Why `instances_status` and not `instance_status`?

Comment (xh3b4sd) [reply service/collector/ec2instances.go]: I was pluralizing all the way through. The singular here would also work. Did not think anything special here.

Comment (marians) [reply service/collector/ec2instances.go]: Changed to `instances_status`

File: service/collector/ec2instances.go (hunk 3 above)
Comment (marians) [reply service/collector/ec2instances.go]: Changed

File: service/collector/ec2instances.go (hunk 4 above)
Comment (marians) [reply service/collector/ec2instances.go]: Done

File: service/collector/ec2instances.go (hunk 5 above)
Comment (marians) [reply service/collector/ec2instances.go]: Going with singular

File: service/collector/ec2instances.go (hunk 7 above)
Comment (marians) [reply service/collector/ec2instances.go]: Changed

**Comment by marians** [comment]
Currently I am seeing these errors on gauss:

```
{"caller":"github.com/giantswarm/aws-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:77","level":"error","message":"failed collecting metrics","stack":"[{/go/src/github.com/giantswarm/aws-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:77: } {/go/src/github.com/giantswarm/aws-operator/service/collector/ec2_instances.go:110: } {/go/src/github.com/giantswarm/aws-operator/service/collector/ec2_instances.go:101: } {/go/src/github.com/giantswarm/aws-operator/service/collector/ec2_instances.go:145: } {UnauthorizedOperation: You are not authorized to perform this operation.\n\tstatus code: 403, request id: df90346c-fca0-4d5f-8b02-590515f62ff5}]","time":"2018-11-26T09:56:52.979965+00:00"}

{"caller":"github.com/giantswarm/aws-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:77","level":"error","message":"failed collecting metrics","stack":"[{/go/src/github.com/giantswarm/aws-operator/vendor/github.com/giantswarm/exporterkit/collector/set.go:77: } {/go/src/github.com/giantswarm/aws-operator/service/collector/asg.go:114: } {/go/src/github.com/giantswarm/aws-operator/service/collector/asg.go:105: } {/go/src/github.com/giantswarm/aws-operator/service/collector/asg.go:143: } {AccessDenied: User: arn:aws:iam::084190472784:user/gauss-aws-operator-host is not authorized to perform: autoscaling:DescribeAutoScalingGroups\n\tstatus code: 403, request id: bcfd52a6-f161-11e8-b85a-35ee777f6781}]","time":"2018-11-26T09:57:52.069334+00:00"}
```

I'll try to update the permissions for `arn:aws:iam::084190472784:user/gauss-aws-operator-host`

**Comment by marians** [comment]
Thanks for the reviews! <3

**Pull request merged**